controllers/ibmpowervsimage_controller.go

```diff
@@ -0,0 +1,172 @@
+
+// Reconcile implements controller runtime Reconciler interface and handles reconcileation logic for IBMPowerVSImage.
+func (r *IBMPowerVSImageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := r.Log.WithValues("ibmpowervsimage", req.NamespacedName)
+
+	ibmImage := &v1beta1.IBMPowerVSImage{}
+	err := r.Get(ctx, req.NamespacedName, ibmImage)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if err := ctrl.SetControllerReference(ibmImage, ibmImage, r.Scheme); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the Image.
+	image, err := util.GetOwnerImage(ctx, r.Client, ibmImage.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if image == nil {
+		log.Info("Image Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	// Create the scope.
+	imageScope, err := scope.NewPowerVSImageScope(scope.PowerVSImageScopeParams{
+		Client:          r.Client,
+		Logger:          log,
+		IBMPowerVSImage: ibmImage,
+	})
+
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := imageScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !ibmImage.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(imageScope)
+	}
+
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+	return r.reconcile(ctx, imageScope)
+}
+
+func (r *IBMPowerVSImageReconciler) reconcile(ctx context.Context, imageScope *scope.PowerVSImageScope) (ctrl.Result, error) {
+	controllerutil.AddFinalizer(imageScope.IBMPowerVSImage, v1beta1.IBMPowerVSImageFinalizer)
+
+	lastJob, err := imageScope.GetImportJob()
+	if err != nil {
+		return ctrl.Result{}, err
+	} else if lastJob != nil {
+		if *lastJob.Status.State != "completed" {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	img, err := r.getOrCreate(imageScope)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to reconcile Image for IBMPowerVSImage %s/%s", imageScope.IBMPowerVSImage.Namespace, imageScope.IBMPowerVSImage.Name)
+	}
+
+	return reconcileImage(img, imageScope)
+}
+
+func reconcileImage(img *models.ImageReference, imageScope *scope.PowerVSImageScope) (_ ctrl.Result, reterr error) {
+	if img != nil {
+		image, err := imageScope.IBMPowerVSClient.ImageClient.Get(*img.ImageID)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		imageScope.IBMPowerVSImage.Status.ImageID = *image.ImageID
+		imageScope.IBMPowerVSImage.Status.ImageState = image.State
+		if imageScope.IBMPowerVSImage.Status.ImageState == "active" {
+			imageScope.IBMPowerVSImage.Status.Ready = true
+		}
+		imageScope.Info(*img.ImageID)
+	} else {
+		imageScope.IBMPowerVSImage.Status.ImageState = "pending"
+		imageScope.IBMPowerVSImage.Status.ImageID = "pending"
```

Comment:
prefer leaving these untouched if no info found instead of adding with some dummy data

Reply:
@mkumatag I had done this as whenever the image is deleted from PowerVS instance, without actually deleting the IBMPowerVSImage resource the reconciler logic should be able to update the status of image back to pending (otherwise it will still be shown as active).

Reply:
In that case we may we should just flush the state instead of making that pending, these field should be one to one mapping with the state of the actual resource in the powervs infra

Reply:
and also any data will be useless if the resource is not in `Ready` state.